### PR TITLE
Implemeted API to update code snippet question

### PR DESCRIPTION
**Pull request by Deepankar-promact**
**Fixed Issues**

#21476

**Files Changed**

QuestionController.cs - Added controller method to update Question
IQuestionRepository.cs - Added function definition of UpdateCodeSnippetQuestionAsync
QuestionRepository.cs - Added implementation of function UpdateCodeSnippetQuestionAsync

**Checks**

- [x] Naming Conventions
- [x] Server side code comments (proper English, grammar and no spelling mistakes)
- [ ] Client side code comments (proper English, grammar and no spelling mistakes)
- [x] Unused variables, methods, blank spaces and name spaces. 
- [x] Optimal Code and code formatting
- [x] Commit History is proper, no merge is done. 
- [x] Commit/pull request messages should be proper to justify your feature
- [ ] All test cases are executed provided by tester.

**Known Issues**

- No issue known

**Comments**

NA

### CODE REVIEW

File: Trappist/src/Promact.Trappist.Repository/Questions/QuestionRepository.cs [hunk 1]
```diff
@@ -109,5 +100,49 @@ public QuestionRepository(TrappistDbContext dbContext)
...
+                updatedCodeSnippetQuestion.Question = updatedQuestion;
+                _dbContext.CodeSnippetQuestion.Update(updatedCodeSnippetQuestion);
+                await _dbContext.SaveChangesAsync();
+
+                //Handling many to many relationship entity
+                //Remove all the mapping between CodeSnippetQuestion and CodingLanguage
+                var mappingToRemove = await _dbContext.QuestionLanguageMapping.Where(x => x.QuestionId == updatedCodeSnippetQuestion.Id).ToListAsync();
+                _dbContext.QuestionLanguageMapping.RemoveRange(mappingToRemove);
+
+                var updatedLanguageList = questionAC.CodeSnippetQuestion.LanguageList;
```
Comment (chintans) [inline Trappist/src/Promact.Trappist.Repository/Questions/QuestionRepository.cs]: MAJOR Remove this unused "updatedLanguageList" local variable. [rule](http://server12:9005/coding_rules#rule_key=csharpsquid%3AS1481)

**Comment by chintans** [comment]
SonarQube analysis reported 1 issue
* ![MAJOR][MAJOR] 1 major

Watch the comments in this conversation to review them.

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'

File: Trappist/src/Promact.Trappist.Core/Controllers/QuestionController.cs [hunk 2]
```diff
@@ -72,6 +73,40 @@ public QuestionController(IQuestionRepository questionsRepository, UserManager<A
...
+        /// <summary>
+        /// Updates an existing Question in the database
+        /// </summary>
+        /// <param name="id">Id of the question to update</param>
+        /// <param name="questionAC">QuestionAC class object</param>
+        /// <returns></returns>
+        [HttpPut("{id}")]
+        public async Task<IActionResult> UpdateQuestion(int id, [FromBody] QuestionAC questionAC)
+        {
+            if (questionAC == null || id == 0 || !ModelState.IsValid)
```
Comment (Krunal-Promact) [inline Trappist/src/Promact.Trappist.Core/Controllers/QuestionController.cs]: if id = 0 then it should return 404 not 400

File: Trappist/src/Promact.Trappist.Core/Controllers/QuestionController.cs [hunk 3]
```diff
@@ -72,6 +73,40 @@ public QuestionController(IQuestionRepository questionsRepository, UserManager<A
...
+            {
+                return BadRequest();
+            }
+
+            try
+            {
+                if (questionAC.Question.QuestionType == QuestionType.Programming)
+                {
+                    await _questionsRepository.UpdateCodeSnippetQuestionAsync(id, questionAC, "");
```
Comment (Krunal-Promact) [inline Trappist/src/Promact.Trappist.Core/Controllers/QuestionController.cs]: what is the meaning of passing an empty string as an argument?

File: Trappist/src/Promact.Trappist.Core/Controllers/QuestionController.cs [hunk 4]
```diff
@@ -72,6 +73,40 @@ public QuestionController(IQuestionRepository questionsRepository, UserManager<A
...
+                }
+                else
+                {
+                    //To-Do Add call to update method for single/multiple type question
+                }
+            }
+            catch (KeyNotFoundException)
```
Comment (Krunal-Promact) [inline Trappist/src/Promact.Trappist.Core/Controllers/QuestionController.cs]: Please justify the try-catch block here.

File: Trappist/src/Promact.Trappist.Core/Controllers/QuestionController.cs [hunk 5]
```diff
@@ -72,6 +73,39 @@ public QuestionController(IQuestionRepository questionsRepository, UserManager<A
...
+        /// <returns></returns>
+        [HttpPut("{id}")]
+        public async Task<IActionResult> UpdateQuestionAsync(int id, [FromBody] QuestionAC questionAC)
+        {
+            if (questionAC == null || !ModelState.IsValid)
+            {
+                return BadRequest();
+            }
+
+            if(!await _questionsRepository.QuestionExist(id))
```
Comment (Krunal-Promact) [inline Trappist/src/Promact.Trappist.Core/Controllers/QuestionController.cs]: method name should be IsQuestionExist()?

File: Trappist/src/Promact.Trappist.Repository/Questions/QuestionRepository.cs [hunk 6]
```diff
@@ -109,5 +100,47 @@ public QuestionRepository(TrappistDbContext dbContext)
             });
             return languageNameList;
         }
+
+        public async Task<bool> QuestionExist(int questionId)
+        {
+            return await _dbContext.Question.Where(x => x.Id == questionId).AnyAsync();
```
Comment (Krunal-Promact) [inline Trappist/src/Promact.Trappist.Repository/Questions/QuestionRepository.cs]: make it efficient.

File: Trappist/src/Promact.Trappist.Repository/Questions/QuestionRepository.cs [hunk 7]
```diff
@@ -109,5 +100,47 @@ public QuestionRepository(TrappistDbContext dbContext)
...
+            //Remove all the mapping between CodeSnippetQuestion and CodingLanguage
+            var mappingToRemove = await _dbContext.QuestionLanguageMapping.Where(x => x.QuestionId == updatedCodeSnippetQuestion.Id).ToListAsync();
+            _dbContext.QuestionLanguageMapping.RemoveRange(mappingToRemove);
+
+            var questionLanguageMapping = new List<QuestionLanguageMapping>();
+
+            //Map language to codeSnippetQuestion
+            foreach (var language in questionAC.CodeSnippetQuestion.LanguageList)
+            {
+                var languageId = await _dbContext.CodingLanguage.Where(x => x.Language == language).Select(x => x.Id).FirstOrDefaultAsync();
```
Comment (Krunal-Promact) [inline Trappist/src/Promact.Trappist.Repository/Questions/QuestionRepository.cs]: var languageId = await _dbContext.CodingLanguage.FirstOrDefaultAsync(x => x.Language == language)?.Id;

Comment (Deepankar-promact) [inline Trappist/src/Promact.Trappist.Repository/Questions/QuestionRepository.cs]: I will update it once you approve the same line in bug fix #21519 i.e PR #217

File: Trappist/src/Promact.Trappist.Repository/Questions/QuestionRepository.cs [hunk 8]
```diff
@@ -101,6 +101,48 @@ public QuestionRepository(TrappistDbContext dbContext)
...
+
+        public async Task UpdateCodeSnippetQuestionAsync(int questionId, QuestionAC questionAC, string userId)
+        {
+            var updatedQuestion = await _dbContext.Question.FirstOrDefaultAsync(x => x.Id == questionId);
+            var updatedCodeSnippetQuestion = await _dbContext.CodeSnippetQuestion.FirstOrDefaultAsync(x => x.Id == questionId);
+
+            Mapper.Map(questionAC.Question, updatedQuestion);
+            Mapper.Map(questionAC.CodeSnippetQuestion, updatedCodeSnippetQuestion);
+
+            updatedQuestion.Id = questionId;
```
Comment (Krunal-Promact) [inline Trappist/src/Promact.Trappist.Repository/Questions/QuestionRepository.cs]: No need to modify primary key.

File: Trappist/src/Promact.Trappist.Repository/Questions/QuestionRepository.cs [hunk 9]
```diff
@@ -101,6 +101,48 @@ public QuestionRepository(TrappistDbContext dbContext)
...
+            updatedQuestion.UpdatedByUserId = userId;
+            updatedCodeSnippetQuestion.Id = questionId;
```
Comment (Krunal-Promact) [inline Trappist/src/Promact.Trappist.Repository/Questions/QuestionRepository.cs]: same as above

File: Trappist/src/Promact.Trappist.Repository/Questions/QuestionRepository.cs [hunk 10]
```diff
@@ -101,6 +101,46 @@ public QuestionRepository(TrappistDbContext dbContext)
...
+            Mapper.Map(questionAC.Question, updatedQuestion);
+            Mapper.Map(questionAC.CodeSnippetQuestion, updatedCodeSnippetQuestion);
+            
+            updatedQuestion.UpdatedByUserId = userId;
+            updatedCodeSnippetQuestion.Question = updatedQuestion;
+
+            _dbContext.Question.Update(updatedQuestion);
+            await _dbContext.SaveChangesAsync();
+            
+            _dbContext.CodeSnippetQuestion.Update(updatedCodeSnippetQuestion);
```
Comment (Krunal-Promact) [inline Trappist/src/Promact.Trappist.Repository/Questions/QuestionRepository.cs]: you have not used transaction here. Any specific reason?

File: Trappist/src/Promact.Trappist.Repository/Questions/QuestionRepository.cs [hunk 11]
```diff
@@ -101,6 +101,46 @@ public QuestionRepository(TrappistDbContext dbContext)
...
+            await _dbContext.SaveChangesAsync();
+
+            //Handling many to many relationship entity
+            //Remove all the mapping between CodeSnippetQuestion and CodingLanguage
+            var mappingToRemove = await _dbContext.QuestionLanguageMapping.Where(x => x.QuestionId == updatedQuestion.CodeSnippetQuestion.Id).ToListAsync();
```
Comment (Krunal-Promact) [inline Trappist/src/Promact.Trappist.Repository/Questions/QuestionRepository.cs]: codingLanguagesToRemove

Comment (Krunal-Promact) [inline Trappist/src/Promact.Trappist.Repository/Questions/QuestionRepository.cs]: variable name is not yet change

File: Trappist/src/Promact.Trappist.Repository/Questions/QuestionRepository.cs [hunk 12]
```diff
@@ -101,6 +101,50 @@ public QuestionRepository(TrappistDbContext dbContext)
             });
             return languageNameList;
         }
+        
+
+        public async Task UpdateCodeSnippetQuestionAsync(int questionId, QuestionAC questionAC, string userId)
+        {
+            var updatedQuestion = await _dbContext.Question.FirstOrDefaultAsync(x => x.Id == questionId);
```
Comment (Krunal-Promact) [inline Trappist/src/Promact.Trappist.Repository/Questions/QuestionRepository.cs]: Please use find method.

File: Trappist/src/Promact.Trappist.Repository/Questions/QuestionRepository.cs [hunk 13]
```diff
@@ -101,6 +101,50 @@ public QuestionRepository(TrappistDbContext dbContext)
...
+            var updatedCodeSnippetQuestion = await _dbContext.CodeSnippetQuestion.FirstOrDefaultAsync(x => x.Id == questionId);
```
Comment (Krunal-Promact) [inline Trappist/src/Promact.Trappist.Repository/Questions/QuestionRepository.cs]: same as above

File: Trappist/src/Promact.Trappist.Test/Questions/QuestionsRepositoryTest.cs [hunk 14]
```diff
@@ -196,21 +199,59 @@ public QuestionsRepositoryTest(Bootstrap bootstrap) : base(bootstrap)
...
+            var updatedQuestion = await _trappistDbContext
+                .Question
+                .Include(x => x.CodeSnippetQuestion)
+                .ThenInclude(x => x.QuestionLanguangeMapping)
+                .FirstOrDefaultAsync(x => x.Id == question.Id);
+
+            Assert.True(string.Equals(updatedQuestion.QuestionDetail, codingQuestion.Question.QuestionDetail, StringComparison.CurrentCultureIgnoreCase));
+            Assert.True(updatedQuestion.DifficultyLevel == codingQuestion.Question.DifficultyLevel);
+            Assert.True(updatedQuestion.CodeSnippetQuestion.CheckCodeComplexity == codingQuestion.CodeSnippetQuestion.CheckCodeComplexity);
+            Assert.True(updatedQuestion.CodeSnippetQuestion.CheckTimeComplexity == codingQuestion.CodeSnippetQuestion.CheckTimeComplexity);
```
Comment (Krunal-Promact) [inline Trappist/src/Promact.Trappist.Test/Questions/QuestionsRepositoryTest.cs]: same as above. No difference found.
Please check all the properties and objects that are being updated.

**Pull request merged**